File: RuyiSDK/RuyiNet/Service/Leaderboard/RuyiNetLeaderboardPage.cs
using System.Collections.Generic;

namespace Ruyi.SDK.Online
{
    /// <summary>
    /// Represents a single page retrieved from a leaderboard.
    /// </summary>
    public class RuyiNetLeaderboardPage
    {
        /// <summary>
        /// Construct from a Get Global Page response.
        /// </summary>
        /// <param name="data">The data from the response</param>
        public RuyiNetLeaderboardPage(RuyiNetGetGlobalLeaderboardPageResponse.Data data)
        {
            Entries = new List<RuyiNetLeaderboardEntry>();
            foreach (var i in data.leaderboard)
            {
                Entries.Add(new RuyiNetLeaderboardEntry(i));
            }

            MoreAfter = data.moreAfter;
            MoreBefore = data.moreBefore;
            TimeBeforeReset = data.timeBeforeReset;
            ServerTime = data.server_time;
        }

        /// <summary>
        /// Construct from a Get Global Page response.
        /// </summary>
        /// <param name="data">The data from the response</param>
        public RuyiNetLeaderboardPage(RuyiNetGetGroupSocialLeaderboardResponse.Data data)
        {
            Entries = new List<RuyiNetLeaderboardEntry>();
            foreach (var i in data.leaderboard)
            {
                Entries.Add(new RuyiNetLeaderboardEntry(i));
            }

            TimeBeforeReset = data.timeBeforeReset;
            ServerTime = data.server_time;
        }

        /// <summary>
        /// Construct from a Get Social Leaderboard response.
        /// </summary>
        /// <param name="data">The data from the response</param>
        public RuyiNetLeaderboardPage(RuyiNetGetSocialLeaderboardResponse.Data data)
        {
            Entries = new List<RuyiNetLeaderboardEntry>();
            foreach (var i in data.social_leaderboard)
            {
                Entries.Add(new RuyiNetLeaderboardEntry(i));
            }

            TimeBeforeReset = data.timeBeforeReset;
            ServerTime = data.server_time;
        }

        /// <summary>
        /// The list of entries for this leaderboard page.
        /// </summary>
        public List<RuyiNetLeaderboardEntry> Entries { get; private set; }

        /// <summary>
        /// True if there are more entries after this page.
        /// </summary>
        public bool MoreAfter { get; private set; }

        /// <summary>
        /// True if there are more entries before this page.
        /// </summary>
        public bool MoreBefore { get; private set;  }

        /// <summary>
        /// How long before the next time this leaderboard is reset.
        /// </summary>
        public int TimeBeforeReset { get; private set; }

        /// <summary>
        /// The server time when this leaderboard was retrieved (UNIX timestamp).
        /// </summary>
        public long ServerTime { get; private set; }
    }
}

File: RuyiSDKCpp/RuyiNet/Service/Leaderboard/RuyiNetLeaderboardPage.cpp
#include "RuyiNetLeaderboardPage.h"

namespace Ruyi { namespace SDK { namespace Online {

	RuyiNetLeaderboardPage::RuyiNetLeaderboardPage() {}

	RuyiNetLeaderboardPage::RuyiNetLeaderboardPage(RuyiNetGetGlobalLeaderboardPageResponse::Data& data)
	{
		std::for_each(data.leaderboard.begin(), data.leaderboard.end(), [&](RuyiNetGetGlobalLeaderboardPageResponse::Data::LeaderboardEntry& entry)
		{
			Entries.push_back(new RuyiNetLeaderboardEntry(entry));
		});

		MoreAfter = data.moreAfter;
		MoreBefore = data.moreBefore;
		TimeBeforeReset = data.timeBeforeReset;
		ServerTime = data.server_time;
	}

	RuyiNetLeaderboardPage::RuyiNetLeaderboardPage(RuyiNetGetGroupSocialLeaderboardResponse::Data& data)
	{
		std::for_each(data.leaderboard.begin(), data.leaderboard.end(), [&](RuyiNetGetGroupSocialLeaderboardResponse::Data::LeaderboardEntry& entry)
		{
			Entries.push_back(new RuyiNetLeaderboardEntry(entry));
		});

		TimeBeforeReset = data.timeBeforeReset;
		ServerTime = data.server_time;
	}

	RuyiNetLeaderboardPage::RuyiNetLeaderboardPage(RuyiNetGetSocialLeaderboardResponse::Data& data)
	{
		std::for_each(data.social_leaderboard.begin(), data.social_leaderboard.end(), [&](RuyiNetGetSocialLeaderboardResponse::Data::LeaderboardEntry& entry)
		{
			Entries.push_back(new RuyiNetLeaderboardEntry(entry));
		});

		TimeBeforeReset = data.timeBeforeReset;
		ServerTime = data.server_time;
	}

	void RuyiNetLeaderboardPage::GetData(RuyiNetGetGlobalLeaderboardPageResponse::Data& data)
	{
		std::for_each(data.leaderboard.begin(), data.leaderboard.end(), [&](RuyiNetGetGlobalLeaderboardPageResponse::Data::LeaderboardEntry& entry)
		{
			Entries.push_back(new RuyiNetLeaderboardEntry(entry));
		});

		MoreAfter = data.moreAfter;
		MoreBefore = data.moreBefore;
		TimeBeforeReset = data.timeBeforeReset;
		ServerTime = data.server_time;
	}

	void RuyiNetLeaderboardPage::GetData(RuyiNetGetGroupSocialLeaderboardResponse::Data& data)
	{
		std::for_each(data.leaderboard.begin(), data.leaderboard.end(), [&](RuyiNetGetGroupSocialLeaderboardResponse::Data::LeaderboardEntry& entry)
		{
			Entries.push_back(new RuyiNetLeaderboardEntry(entry));
		});

		TimeBeforeReset = data.timeBeforeReset;
		ServerTime = data.server_time;
	}

	void RuyiNetLeaderboardPage::GetData(RuyiNetGetSocialLeaderboardResponse::Data& data)
	{
		std::for_each(data.social_leaderboard.begin(), data.social_leaderboard.end(), [&](RuyiNetGetSocialLeaderboardResponse::Data::LeaderboardEntry& entry)
		{
			Entries.push_back(new RuyiNetLeaderboardEntry(entry));
		});

		TimeBeforeReset = data.timeBeforeReset;
		ServerTime = data.server_time;
	}

	std::vector<RuyiNetLeaderboardEntry*>& RuyiNetLeaderboardPage::GetEntries() { return Entries; }
	bool RuyiNetLeaderboardPage::GetMoreAfter() { return MoreAfter; }
	bool RuyiNetLeaderboardPage::GetMoreBefore() { return MoreBefore; }
	int RuyiNetLeaderboardPage::GetTimeBeforeReset() { return TimeBeforeReset; }
	long RuyiNetLeaderboardPage::GetServerTime() { return ServerTime; }

}}}